### PR TITLE
Handle changes to ID mapping in images when possible

### PR DESCRIPTION
**Pull request by nalind**
Currently, if we want to create two containers based on a single image, and have those containers share a single ID mapping configuration that differs from the one we're using for the image's layers, we create the layers for each of the containers and use `chown()` to fix up the permissions.

This patch set modifies the storage interface so that when the image is in a read-write store, we instead create a new copy of the "top" layer for the image that uses the desired mappings, track that new layer using a new field in the `Image` structure, and base the container's layer on that.  When we go to create a second container with identical mappings, we'll reuse that layer as the base.  (When the image is in a read-only store, we continue to "hide" the remapping in the container's layer, as before.)

### CODE REVIEW

File: store.go [hunk 1]
```diff
@@ -1730,11 +1819,26 @@ func (s *store) DeleteImage(id string, commit bool) (layers []string, err error)
 					}
 				}
 			}
+			if layer == image.TopLayer {
+				for _, mappedLayer := range image.MappedTopLayers {
+					if childList, ok := childrenByParent[mappedLayer]; ok && childList != nil {
+						children := *childList
+						for _, child := range children {
+							if child != lastRemoved {
+								otherRefs++
```
Comment (rhatdan) [inline store.go]: Seems like this should be a function and just return at this point.

**Comment by rhatdan** [comment]
@mtrmac @mheon @mrunalp PTAL

File: store.go (hunk 1 above)
Comment (nalind) [reply store.go]: Made it a function.

File: store.go [hunk 2]
```diff
@@ -949,6 +951,101 @@ func (s *store) CreateImage(id string, names []string, layer, metadata string, o
...
+			}
+		}
+	}
+	if layer == nil || (layer.Parent != "" && parentLayer == nil) {
+		return nil, ErrLayerUnknown
+	}
+	// If the top layer's mappings don't match the ones we want, and it's in an image store
+	// that lets us edit image metadata...
+	if !layerMatchesMappingOptions(layer, options) && readWrite {
+		if istore, ok := ristore.(*imageStore); ok && (layer.Parent == "" || parentLayer != nil) {
```
[the remark this answers is not in the document]
Comment (nalind) [reply store.go]: Ah, right.  Removing it.

File: store.go [hunk 3]
```diff
@@ -1714,27 +1803,44 @@ func (s *store) DeleteImage(id string, commit bool) (layers []string, err error)
...
 				parent = l.Parent
 			}
-			otherRefs := 0
-			if childList, ok := childrenByParent[layer]; ok && childList != nil {
-				children := *childList
-				for _, child := range children {
-					if child != lastRemoved {
-						otherRefs++
+			hasOtherRefs := func() bool {
+				if childList, ok := childrenByParent[layer]; ok && childList != nil {
```
[the remark this answers is not in the document]
Comment (nalind) [reply store.go]: We can factor it, yeah.

File: store.go [hunk 4]
```diff
@@ -949,6 +951,101 @@ func (s *store) CreateImage(id string, names []string, layer, metadata string, o
...
+					if cParentLayer, err = store.Get(cLayer.Parent); err != nil {
+						continue
+					}
+				}
+				// If the layer matches the desired mappings, it's a perfect match.
+				if layerMatchesMappingOptions(cLayer, options) {
+					layer = cLayer
+					parentLayer = cParentLayer
+					layerHomeStore = store
+					break
```
[the remark this answers is not in the document]
Comment (nalind) [reply store.go]: Right on both counts - if the layer we're looking at already has the mappings we're looking for, we want to use it as-is.  Making those changes.

File: store.go [hunk 5]
```diff
@@ -949,6 +951,101 @@ func (s *store) CreateImage(id string, names []string, layer, metadata string, o
...
+				// Record the first one that we found, even if it's not ideal.
+				if layer == nil {
+					layer = cLayer
+					parentLayer = cParentLayer
+					layerHomeStore = store
+				}
+			}
+		}
+	}
+	if layer == nil || (layer.Parent != "" && parentLayer == nil) {
```
[the remark this answers is not in the document]
Comment (nalind) [reply store.go]: Hmm, it shouldn't happen now, anyway.  Dropping that check.

**Comment by nalind** [comment]
[containers/storage: trade disk space for userns chown speed](https://trello.com/c/O7CyDsf7/604-containers-storage-trade-disk-space-for-userns-chown-speed)

**Comment by nalind** [comment]
Added `ImageSize()` and `ContainerSize()` methods, because we just made the image size calculation somewhat more complicated than it used to be.

**Comment by rhatdan** [comment]
@giuseppe This is the PR we want to start testing with podman.

**Comment by giuseppe** [comment]
podman is much faster with this change.

The image is already pulled and owned by root, so no pull is required but it must be copied and chown'ed for each different mapping:

```console
# time bin/podman run --rm -it --uidmap='0:100000:100000' --gidmap='0:100000:100000' fedora true 3.92user 1.95system 0:05.77elapsed 101%CPU (0avgtext+0avgdata 72244maxresident)k
0inputs+545848outputs (0major+34757minor)pagefaults 0swaps

# time bin/podman run --rm -it --uidmap='0:100000:100000' --gidmap='0:100000:100000' fedora true
0.18user 0.17system 0:00.66elapsed 53%CPU (0avgtext+0avgdata 31748maxresident)k
0inputs+832outputs (0major+19659minor)pagefaults 0swaps

# time bin/podman run --rm -it --uidmap='0:100001:100000' --gidmap='0:100001:100000' fedora true
4.03user 2.08system 0:05.44elapsed 112%CPU (0avgtext+0avgdata 77604maxresident)k
536inputs+545848outputs (0major+36047minor)pagefaults 0swaps

# time bin/podman run --rm -it --uidmap='0:100001:100000' --gidmap='0:100001:100000' fedora true
0.14user 0.15system 0:00.58elapsed 52%CPU (0avgtext+0avgdata 31712maxresident)k
0inputs+832outputs (0major+19539minor)pagefaults 0swaps
```

**Comment by TomSweeneyRedHat** [comment]
LGTM, but definitely would like some other head nods.

**Comment by rhatdan** [comment]
@giuseppe @mtrmac PTAL

**Comment by giuseppe** [comment]
I am nor really familiar with the codebase, but LGTM

**Comment by giuseppe** [comment]
they seem to work well, "podman images" shows the image just once as IMHO should be, and "podman rmi" removes all the copies

**Comment by rhatdan** [comment]
Ok lets get this merged and then we can vendor it into podman.

File: store.go [hunk 6]
```diff
@@ -949,6 +959,106 @@ func (s *store) CreateImage(id string, names []string, layer, metadata string, o
...
+		}
+		if options.HostGIDMapping && len(layer.GIDMap) != 0 {
+			return false
+		}
+		// If we don't care about the mapping, it's fine.
+		if len(options.UIDMap) == 0 && len(options.GIDMap) == 0 {
+			return true
+		}
+		// Compare the maps.
+		return reflect.DeepEqual(layer.UIDMap, options.UIDMap) && reflect.DeepEqual(layer.GIDMap, options.GIDMap)
```
Comment (giuseppe) [inline store.go]: one question I just got now, do we get two different images if mappings like `0:1:1000` and `0:1:2000` are used?

**Pull request merged**